src/Control/Distributed/Process/Extras/Internal/Primitives.hs

```diff
@@ -70,44 +70,38 @@ import Control.Distributed.Process.Extras.Internal.Types
   , ExitReason(ExitOther)
   , whereisRemote
   )
-import Control.Monad (void)
-import Data.Maybe (isJust, fromJust)
+import Control.Monad (void, (>=>), replicateM_)
+import Control.Monad.Catch (bracket)
+import Data.Maybe (isJust, fromJust, isJust)
+import Data.Foldable (traverse_)
 
 -- utility
 
 -- | Monitor any @Resolvable@ object.
---
 monitor :: Resolvable a => a -> Process (Maybe MonitorRef)
-monitor addr = do
-  mPid <- resolve addr
-  case mPid of
-    Nothing -> return Nothing
-    Just p  -> return . Just =<< P.monitor p
+monitor = resolve >=> traverse P.monitor
 
+-- | Wait until @Resolvable@ object will exit. Return immediately
+-- if object can't be resolved.
 awaitExit :: Resolvable a => a -> Process ()
-awaitExit addr = do
-  mPid <- resolve addr
-  case mPid of
-    Nothing -> return ()
-    Just p  -> do
-      mRef <- P.monitor p
-      receiveWait [
-          matchIf (\(ProcessMonitorNotification r p' _) -> r == mRef && p == p')
-                  (\_ -> return ())
-        ]
+awaitExit = resolve >=> traverse_ await where
+  await pid = withMonitor pid $ receiveWait
+      [ matchIf (\(ProcessMonitorNotification _ p _) -> p == pid)
```

Comment:
This line isn't okay - it could be the wrong monitor signal we pick up, because we don't validate the mref. We could either wait for `withMonitorRef` to land in distributed-process and raise our dependency, or I think we should stick to using the monitor ref as before